### PR TITLE
chore: perform authentication check in react router [DET-3929]

### PR DESCRIPTION
**Pull request by hamidzr**
## Test Plan

- [x] tested on a slow 3g network

## Commentary (optional)

~If we want to forgo having pages that do not require auth (other than login) one option is to move the auth check up a level and into app.tsx. there we can go with a different code path before engaging the router. without doing so the router has to send the user somewhere while it is waiting to check the auth status with server and that somewhere with this PR is the target component. going anywhere else would have the effect of seeing a route change~

## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

**Comment by hamidzr** [comment]
useCheck in the sign in page seems to be necessary for automatically logging back in.

File: webui/react/src/contexts/Auth.tsx [hunk 1]
```diff
@@ -8,6 +9,8 @@ enum ActionType {
   Set,
 }
 
+export const AUTH_COOKIE_KEY = 'auth';
```
Comment (hkang1) [inline webui/react/src/contexts/Auth.tsx:12]: 👍 nice

File: webui/react/src/components/Router.tsx [hunk 2]
```diff
@@ -12,13 +14,27 @@ interface Props {
...
+  }, [ auth.isAuthenticated, setUI ]);
 
   return (
     <Switch>
       {props.routes.map(config => {
         const { component, ...route } = config;
 
         if (route.needAuth && !auth.isAuthenticated) {
+          // Do not mount login page until auth is checked.
+          if (!auth.checked) return <Route key={route.id} {...route} />;
```
Comment (hkang1) [inline webui/react/src/components/Router.tsx:37]: This is sweet! So if it doesn't have `render` or `component` defined, it'll default to a blank screen?

Comment (hamidzr) [reply webui/react/src/components/Router.tsx:37]: I initially went with an empty fragment but this worked. not necessarily a blank screen but no component for that portion of the app? I couldn't find a good reference to this in https://knowbody.github.io/react-router-docs/api/Route.html

the goal is to
- not render a component if needs auth and we're not authenticated
- avoid visual flickers in the UI

**Pull request merged**